### DISCUSSION TITLE
Recreate the C4 Model example using Gaphor

### DISCUSSION THREAD

**Discussion by marcus-adamski** [gaphor/gaphor#3412] [Q&A]
I'm looking to model a new system using Gaphor and C4. However, I'm having difficulties understanding how to use Gaphor (which toolbox elements) are required for the different abstractions (context, container, component, code). 

Would it possible for a knowledgeable person to correctly model just part of the C4 example "Internet Banking System" (for the different abstractions),  using Gaphor and upload the project file. It would really help new users to explore and understand C4 modelling using Gaphor. The C4 "Internet Banking System" example can be found on the home page: https://c4model.com

Thank you.

**Comment by amolenaar** [comment]
Hi @marcus-adamski,

HVing a proper C4 example would be a nice addition. As far as I can tell all C4 elements have been implemented.

Do you want to create one?

**Reply by marcus-adamski** [reply]
Hi @amolenaar - unfortunately, I'm new to both Gaphor and the C4 model. After reading the C4 model site I've got a good idea of how the model works, but I'm confused how to implement it through Gaphor's tool set.

If someone with more experience could recreate even part of the official C4 example so I can download and explore it, that would really help my learning curve.

**Reply by amolenaar** [reply]
@marcus-adamski Please have a look at https://github.com/gaphor/gaphor/pull/3416. It requires the latest, unreleased, code, though.

**Comment by marcus-adamski** [comment]
@amolenaar thank-you for creating the C4 example so quickly. I've managed to build the branch and explored Gaphor and the example, which has given me a great insight on how this hangs together. So for the c4 models, the dependency relationship is used for the context, container & component modules - that's something I wasn't sure about. Quick question - I can see the context diagram->internet banking system is the same element as the container diagram->internet banking system, so I'd expect to be able to double-click on the internet banking system in the context and be taken to the container diagram. This would be useful to following a path through the model, and if there were multiple container diagrams modelling different software systems. The second question I have is, can I create a model using this branch build and expect it to work with the next Gaphor release? Thanks again for your help!

**Comment by amolenaar** [comment]
> ... so I'd expect to be able to double-click on the internet banking system in the context and be taken to the container diagram. This would be useful to following a path through the model, and if there were multiple container diagrams modelling different software systems. 

This doesn't work, since the diagram contains multiple elements. What we can do is:

1. Add the more-detailed diagram next to the component (or inside)
2. The Model browser popup menu shows in which diagrams an element is presented. We can add something similar for the popup menu in the diagram.

Currently, the diagrams are in separate packages. It may make sense to add those to the container/component the diagram is about.

> The second question I have is, can I create a model using this branch build and expect it to work with the next Gaphor release? Thanks again for your help!

Yes, this will be part of the next release. It doesn't load in 2.26.0, since I used a new C4 dependency element, where we can provide a descriptive name and technology to the relationship.

**Comment by danyeaw** [comment]
Hi @marcus-adamski, thanks for the discussion and suggestion. @amolenaar implemented this here: https://github.com/gaphor/gaphor/blob/main/gaphor/templates/c4model-example.gaphor
